io/plugins/eu.esdihumboldt.hale.io.csv.test/src/eu/esdihumboldt/hale/io/csv/writer/CSVInstanceWriterTest.java

```diff
@@ -76,42 +77,112 @@ public void testWriteSimpleSchema() throws Exception {
+				Value.of(sep), null, null, example.getSourceInstances()));
 
 		CSVReader reader = new CSVReader(new FileReader(tmpFile), sep);
 		List<String[]> rows = reader.readAll();
-		//
+
 		reader.close();
 
 		assertEquals("Not enough rows.", numberOfRows, rows.size());
 
+		// Check no. of columns ###
+		int countCols = rows.get(0).length;
+		// If b1 is an empty string in instance1.xml,
+		// the column is not dropped
+		System.out.print("TEST written columns " + Arrays.toString(rows.get(0)));
+		assertEquals("Not enough columns written in the output file.", numberOfColumns, countCols);
+		// The column with no values is dropped when useSchema=false
+
 		// Check header ###
 		Iterator<String[]> row = rows.iterator();
 
 		String[] header = row.next();
-		assertEquals("There are not enough entries.", numberOfEntries, header.length);
+		assertEquals("There are not enough entries.", numberOfColumns, header.length);
 
 		for (int i = 0; i < header.length; i++) {
-			assertTrue("The header of the csv file do not contain all properties.",
+			assertTrue("The header of the csv file does not contain all properties.",
 					propertyNames.contains(header[i]));
 		}
 
 		String[] dataRow = row.next();
 		for (int i = 0; i < dataRow.length; i++) {
-			assertTrue("The first data row of the csv file do not contain all properties.",
+			assertTrue("The first data row of the csv file does not contain all properties.",
+					firstDataRow.contains(dataRow[i]));
+		}
+	}
+
+	/**
+	 * Test - write simple data, without nested properties and useSchema=true -
+	 * order of column persisted from original schema
+	 * 
+	 * @throws Exception , if an error occurs
+	 */
+	@Test
+	public void testWriteSimpleSchemaColOrder() throws Exception {
+
+		TransformationExample example = TransformationExamples
+				.getExample(TransformationExamples.SIMPLE_ASSIGN);
+		// alternative the data could be generated by iterating through the
+		// exemple project's source data:
+		// eu.esdihumboldt.cst.test/src/testdata/simpleassign/instance1.xml
+		String propertyNames = "a1,b1,c1,id";
+		String firstDataRow = "a10,,c10,id0";
+		// header size
+		int numberOfColumns = 4; // in the example t1.xsd: elements + attribute
+		int numberOfRows = 3;
+		char sep = ',';
+
+		File tmpFile = tmpFolder.newFile("csvTestWriteSimpleSchema.csv");
+
+		assertTrue("Csv Export was not successful.", writeCsvToFile(tmpFile, true, true,
+				Value.of(sep), null, null, example.getSourceInstances()));
+
+		CSVReader reader = new CSVReader(new FileReader(tmpFile), sep);
+		List<String[]> rows = reader.readAll();
+
+		reader.close();
+
+		assertEquals("Not enough rows.", numberOfRows, rows.size());
+
+		// Check no. of columns ###
+		int countCols = rows.get(0).length;
+		assertEquals("Not enough columns written in the output file.", numberOfColumns, countCols);
+		// Check the order of the columns ###
+		String[] originalColumns = propertyNames.split(",");
+		int j = 0;
+		for (String s : originalColumns) {
```

Comment:
Since you are not using `s` in the loop nor the order from the originalColumn and only its length, the loop can be modified as: 

```
for (int j = 0; j < originalColumns.length; j++) {
                      
                    }
```

At first looking at the loop, it appears as if you would be using `s` inside it but as it is not the case so I would advise to change the loop to use `originalColumns.length`.

Reply:
I think `s` it is used, look in the assert at line 173:
```
for (String s : originalColumns) {
			assertEquals("The order of the columns is not equal to the original source file", s,
					rows.get(0)[j]);
			j++;
		}
```